packages/flutter_tools/lib/src/android/gradle.dart

```diff
@@ -244,6 +317,74 @@ distributionUrl=https\\://services.gradle.org/distributions/gradle-$gradleVersio
   }
 }
 
+/// Returns true if [targetVersion] is within the range [min] and [max] inclusive.
+bool _isWithinVersionRange(String targetVersion, {String min, String max}) {
+  final Version parsedTargetVersion = Version.parse(targetVersion);
+  return parsedTargetVersion >= Version.parse(min) &&
+      parsedTargetVersion <= Version.parse(max);
+}
+
+const String defaultGradleVersion = '4.10.2';
+
+/// Returns the Gradle version that is required by the given Android Gradle plugin version
+/// by picking the largest compatible version from
+/// https://developer.android.com/studio/releases/gradle-plugin#updating-gradle
+String getGradleVersionFor(String androidPluginVersion) {
+  if (_isWithinVersionRange(androidPluginVersion, min: '1.0.0', max: '1.1.3')) {
+    return '2.3';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '1.2.0', max: '1.3.1')) {
+    return '2.9';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '1.5.0', max: '1.5.0')) {
+    return '2.2.1';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '2.0.0', max: '2.1.2')) {
+    return '2.13';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '2.1.3', max: '2.2.3')) {
+    return '2.14.1';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '2.3.0', max: '2.9.9')) {
+    return '3.3';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '3.0.0', max: '3.0.9')) {
+    return '4.1';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '3.1.0', max: '3.1.9')) {
+    return '4.4';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '3.2.0', max: '3.2.1')) {
+    return '4.6';
+  }
+  if (_isWithinVersionRange(androidPluginVersion, min: '3.3.0', max: '3.3.2')) {
+    return '4.10.2';
+  }
+  return '5.1.1';
```

Comment:
We should cap this at anything under 3.5.0 - the pattern has been that changing the minimum gradle version has resulted in a minor release of the Android plugin based on the linked table. What we want to avoid here is releasing this to stable, having gradle plugin 3.5.0 released that requires Gradle 5.2 or something, and the tool is trying to force you to use Gradle 5.1.1.

That said, it'd be really nice if we could come up with some way to keep this from bitrotting - may be worth checking with the Android gradle team if they have a stable source of this information somewhere that we could consume.

Reply:
We might just want to print some kind of warning/instructions if the version >= 3.5.0

Reply:
We control the plugin version in the template file. This is to _fix_ a mistake in the existing plugins.

Reply:
As written, this fix will stop working when Android Gradle plugin 3.4.0 is released (assuming the ycontinue to use the same conventions)...  right?

Reply:
We set the Android Gradle plugin in project's generated file, which all projects have except plugins. Hence, I added this workaround.

For new plugins, we shouldn't need this code at all because the new generated files explicitly indicate both the Gradle version and the Android Gradle plugin.